sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala

```diff
@@ -683,9 +685,29 @@ abstract class HadoopFsRelation private[sql](
       }
     }
 
-    buildInternalScan(requiredColumns, filters, inputStatuses, broadcastedConf)
+    if (bucketSpec.isDefined && sqlContext.conf.bucketingEnabled()) {
+      val inputFilesWithBucketId = allInputFiles.groupBy(getBucketId)
```

Comment:
I'm still concerned about perf here. This should be O(number of files) and not O(number of files, number of buckets). Getting metadata related things are slow enough as is when the number of files is large.